### DISCUSSION TITLE
如何在多个用户控件中对同一个数据对象进行操作

### DISCUSSION THREAD

**Discussion by jcwn888** [BYJRK/DotNet-Discussions#120] [WPF 开发]
开发一个应用，数据是一个具有多层嵌套的复杂对象（命名为DataObject），使用多个子窗口操作和展示同一个DataObject对象的不同方面，每个子窗口里是一个用户控件（多窗口布局用的是AvalonDock）。
如果在某个子窗口中更改了DataObject，其它子窗口要自动刷新显示新的DataObject。
如何使用MVVM模式实现以上功能？
或者有什么具有以上特性的开源项目推荐？
谢谢大家了（目前只知道怎么处理ViewModel与窗口（或用户控件）一对一的情况）
![image](https://github.com/BYJRK/DotNet-Discussions/assets/32017388/47e624ef-79b9-4b8a-9a75-40451554c99c)

**Comment by CodingOctocat** [comment]
用 MvvmToolkit 的 Messenger 就行了。

MainVm：

```CS
public class MainVm: ObservableRecipient
{
    [ObservableProperty]
    [NotifyPropertyChangedRecipients]
    private DataObject _dataObject;
}
```

各个 子窗口/UserControl 的 Vm：

```CS
public class OtherVm: ObservableRecipient
{
    [ObservableProperty]
    private DataObject _dataObject;

    public void Receive(PropertyChangedMessage<DataObject> message)
    {
        _dataObject = message.NewValue;
    }
}
```

Vm 可以和 View 是 `一对多` 的关系。如果你的 子窗口/UserControl 的逻辑没必要拆成单独的 Vm，那么共用一个 Vm 就行了，不过要注意，你可能需要在依赖注入的时候将这个 Vm 当成单例来使用：`services.AddSingleton<YourVm>();`，不然 DataObject  实例是不一样的。

参考示例：https://learn.microsoft.com/zh-cn/dotnet/communitytoolkit/mvvm/puttingthingstogether

**Comment by ErickWei** [comment]
> 开发一个应用，数据是一个具有多层嵌套的复杂对象（命名为DataObject），使用多个子窗口操作和展示同一个DataObject对象的不同方面，每个子窗口里是一个用户控件（多窗口布局用的是AvalonDock）。 如果在某个子窗口中更改了DataObject，其它子窗口要自动刷新显示新的DataObject。 如何使用MVVM模式实现以上功能？ 或者有什么具有以上特性的开源项目推荐？ 谢谢大家了（目前只知道怎么处理ViewModel与窗口（或用户控件）一对一的情况） ![image](https://private-user-images.githubusercontent.com/32017388/253143083-47e624ef-79b9-4b8a-9a75-40451554c99c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5MTI1MDksIm5iZiI6MTcwMTkxMjIwOSwicGF0aCI6Ii8zMjAxNzM4OC8yNTMxNDMwODMtNDdlNjI0ZWYtNzliOS00YjhhLTlhNzUtNDA0NTE1NTRjOTljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA3VDAxMjMyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzNWY1YTMzMTk3ZTk1ZDBhYjI3OWQ2ODdmMjQ2MjhkMjkwYWRhNjhkYmUxNTI5ZWFiZGQwMWRhY2E2OGRlNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3jCb61Ik4QL9l9EDbb9DKUtDctjFzBfvW274gBdsBcs)

请问您这里使用了什么UI 控件？看上去很好看

**Reply by CodingOctocat** [reply]
[HandyControl](https://github.com/HandyOrg/HandyControl)

**Comment by ErickWei** [comment]
> [HandyControl](https://github.com/HandyOrg/HandyControl)

感觉不像哈  HC 里没有这样的控件

<img width="265" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/62826892/55dc8187-4922-48de-9655-7b5196bfa2ba">

**Reply by CodingOctocat** [reply]
[AvalonDock](https://github.com/Dirkster99/AvalonDock)

**Comment by ErickWei** [comment]
> [AvalonDock](https://github.com/Dirkster99/AvalonDock)
多谢啊，确实是这个
